
%% bare_jrnl.tex
%% V1.4
%% 2012/12/27
%% by Michael Shell
%% see http://www.michaelshell.org/
%% for current contact information.
%%
%% This is a skeleton file demonstrating the use of IEEEtran.cls
%% (requires IEEEtran.cls version 1.8 or later) with an IEEE journal paper.
%%
%% Support sites:
%% http://www.michaelshell.org/tex/ieeetran/
%% http://www.ctan.org/tex-archive/macros/latex/contrib/IEEEtran/
%% and
%% http://www.ieee.org/



% *** Authors should verify (and, if needed, correct) their LaTeX system  ***
% *** with the testflow diagnostic prior to trusting their LaTeX platform ***
% *** with production work. IEEE's font choices can trigger bugs that do  ***
% *** not appear when using other class files.                            ***
% The testflow support page is at:
% http://www.michaelshell.org/tex/testflow/


%%*************************************************************************
%% Legal Notice:
%% This code is offered as-is without any warranty either expressed or
%% implied; without even the implied warranty of MERCHANTABILITY or
%% FITNESS FOR A PARTICULAR PURPOSE! 
%% User assumes all risk.
%% In no event shall IEEE or any contributor to this code be liable for
%% any damages or losses, including, but not limited to, incidental,
%% consequential, or any other damages, resulting from the use or misuse
%% of any information contained here.
%%
%% All comments are the opinions of their respective authors and are not
%% necessarily endorsed by the IEEE.
%%
%% This work is distributed under the LaTeX Project Public License (LPPL)
%% ( http://www.latex-project.org/ ) version 1.3, and may be freely used,
%% distributed and modified. A copy of the LPPL, version 1.3, is included
%% in the base LaTeX documentation of all distributions of LaTeX released
%% 2003/12/01 or later.
%% Retain all contribution notices and credits.
%% ** Modified files should be clearly indicated as such, including  **
%% ** renaming them and changing author support contact information. **
%%
%% File list of work: IEEEtran.cls, IEEEtran_HOWTO.pdf, bare_adv.tex,
%%                    bare_conf.tex, bare_jrnl.tex, bare_jrnl_compsoc.tex,
%%                    bare_jrnl_transmag.tex
%%*************************************************************************

% Note that the a4paper option is mainly intended so that authors in
% countries using A4 can easily print to A4 and see how their papers will
% look in print - the typesetting of the document will not typically be
% affected with changes in paper size (but the bottom and side margins will).
% Use the testflow package mentioned above to verify correct handling of
% both paper sizes by the user's LaTeX system.
%
% Also note that the "draftcls" or "draftclsnofoot", not "draft", option
% should be used if it is desired that the figures are to be displayed in
% draft mode.
%
\documentclass[journal]{IEEEtran}
%
% If IEEEtran.cls has not been installed into the LaTeX system files,
% manually specify the path to it like:
% \documentclass[journal]{../sty/IEEEtran}

%\linespread{2.5}



% Some very useful LaTeX packages include:
% (uncomment the ones you want to load)


% *** MISC UTILITY PACKAGES ***
%
%\usepackage{ifpdf}
% Heiko Oberdiek's ifpdf.sty is very useful if you need conditional
% compilation based on whether the output is pdf or dvi.
% usage:
% \ifpdf
%   % pdf code
% \else
%   % dvi code
% \fi
% The latest version of ifpdf.sty can be obtained from:
% http://www.ctan.org/tex-archive/macros/latex/contrib/oberdiek/
% Also, note that IEEEtran.cls V1.7 and later provides a builtin
% \ifCLASSINFOpdf conditional that works the same way.
% When switching from latex to pdflatex and vice-versa, the compiler may
% have to be run twice to clear warning/error messages.






% *** CITATION PACKAGES ***
%
%\usepackage{cite}
% cite.sty was written by Donald Arseneau
% V1.6 and later of IEEEtran pre-defines the format of the cite.sty package
% \cite{} output to follow that of IEEE. Loading the cite package will
% result in citation numbers being automatically sorted and properly
% "compressed/ranged". e.g., [1], [9], [2], [7], [5], [6] without using
% cite.sty will become [1], [2], [5]--[7], [9] using cite.sty. cite.sty's
% \cite will automatically add leading space, if needed. Use cite.sty's
% noadjust option (cite.sty V3.8 and later) if you want to turn this off
% such as if a citation ever needs to be enclosed in parenthesis.
% cite.sty is already installed on most LaTeX systems. Be sure and use
% version 4.0 (2003-05-27) and later if using hyperref.sty. cite.sty does
% not currently provide for hyperlinked citations.
% The latest version can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/cite/
% The documentation is contained in the cite.sty file itself.






% *** GRAPHICS RELATED PACKAGES ***
%
\ifCLASSINFOpdf
  % \usepackage[pdftex]{graphicx}
  % declare the path(s) where your graphic files are
  % \graphicspath{{../pdf/}{../jpeg/}}
  % and their extensions so you won't have to specify these with
  % every instance of \includegraphics
  % \DeclareGraphicsExtensions{.pdf,.jpeg,.png}
\else
  % or other class option (dvipsone, dvipdf, if not using dvips). graphicx
  % will default to the driver specified in the system graphics.cfg if no
  % driver is specified.
  % \usepackage[dvips]{graphicx}
  % declare the path(s) where your graphic files are
  % \graphicspath{{../eps/}}
  % and their extensions so you won't have to specify these with
  % every instance of \includegraphics
  % \DeclareGraphicsExtensions{.eps}
\fi
% graphicx was written by David Carlisle and Sebastian Rahtz. It is
% required if you want graphics, photos, etc. graphicx.sty is already
% installed on most LaTeX systems. The latest version and documentation
% can be obtained at: 
% http://www.ctan.org/tex-archive/macros/latex/required/graphics/
% Another good source of documentation is "Using Imported Graphics in
% LaTeX2e" by Keith Reckdahl which can be found at:
% http://www.ctan.org/tex-archive/info/epslatex/
%
% latex, and pdflatex in dvi mode, support graphics in encapsulated
% postscript (.eps) format. pdflatex in pdf mode supports graphics
% in .pdf, .jpeg, .png and .mps (metapost) formats. Users should ensure
% that all non-photo figures use a vector format (.eps, .pdf, .mps) and
% not a bitmapped formats (.jpeg, .png). IEEE frowns on bitmapped formats
% which can result in "jaggedy"/blurry rendering of lines and letters as
% well as large increases in file sizes.
%
% You can find documentation about the pdfTeX application at:
% http://www.tug.org/applications/pdftex





% *** MATH PACKAGES ***
%
%\usepackage[cmex10]{amsmath}
% A popular package from the American Mathematical Society that provides
% many useful and powerful commands for dealing with mathematics. If using
% it, be sure to load this package with the cmex10 option to ensure that
% only type 1 fonts will utilized at all point sizes. Without this option,
% it is possible that some math symbols, particularly those within
% footnotes, will be rendered in bitmap form which will result in a
% document that can not be IEEE Xplore compliant!
%
% Also, note that the amsmath package sets \interdisplaylinepenalty to 10000
% thus preventing page breaks from occurring within multiline equations. Use:
%\interdisplaylinepenalty=2500
% after loading amsmath to restore such page breaks as IEEEtran.cls normally
% does. amsmath.sty is already installed on most LaTeX systems. The latest
% version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/required/amslatex/math/





% *** SPECIALIZED LIST PACKAGES ***
%
%\usepackage{algorithmic}
\usepackage{algpseudocode}
% algorithmic.sty was written by Peter Williams and Rogerio Brito.
% This package provides an algorithmic environment fo describing algorithms.
% You can use the algorithmic environment in-text or within a figure
% environment to provide for a floating algorithm. Do NOT use the algorithm
% floating environment provided by algorithm.sty (by the same authors) or
% algorithm2e.sty (by Christophe Fiorio) as IEEE does not use dedicated
% algorithm float types and packages that provide these will not provide
% correct IEEE style captions. The latest version and documentation of
% algorithmic.sty can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/algorithms/
% There is also a support site at:
% http://algorithms.berlios.de/index.html
% Also of interest may be the (relatively newer and more customizable)
% algorithmicx.sty package by Szasz Janos:
% http://www.ctan.org/tex-archive/macros/latex/contrib/algorithmicx/




% *** ALIGNMENT PACKAGES ***
%
%\usepackage{array}
% Frank Mittelbach's and David Carlisle's array.sty patches and improves
% the standard LaTeX2e array and tabular environments to provide better
% appearance and additional user controls. As the default LaTeX2e table
% generation code is lacking to the point of almost being broken with
% respect to the quality of the end results, all users are strongly
% advised to use an enhanced (at the very least that provided by array.sty)
% set of table tools. array.sty is already installed on most systems. The
% latest version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/required/tools/


% IEEEtran contains the IEEEeqnarray family of commands that can be used to
% generate multiline equations as well as matrices, tables, etc., of high
% quality.




% *** SUBFIGURE PACKAGES ***
%\ifCLASSOPTIONcompsoc
%  \usepackage[caption=false,font=normalsize,labelfont=sf,textfont=sf]{subfig}
%\else
%  \usepackage[caption=false,font=footnotesize]{subfig}
%\fi
% subfig.sty, written by Steven Douglas Cochran, is the modern replacement
% for subfigure.sty, the latter of which is no longer maintained and is
% incompatible with some LaTeX packages including fixltx2e. However,
% subfig.sty requires and automatically loads Axel Sommerfeldt's caption.sty
% which will override IEEEtran.cls' handling of captions and this will result
% in non-IEEE style figure/table captions. To prevent this problem, be sure
% and invoke subfig.sty's "caption=false" package option (available since
% subfig.sty version 1.3, 2005/06/28) as this is will preserve IEEEtran.cls
% handling of captions.
% Note that the Computer Society format requires a larger sans serif font
% than the serif footnote size font used in traditional IEEE formatting
% and thus the need to invoke different subfig.sty package options depending
% on whether compsoc mode has been enabled.
%
% The latest version and documentation of subfig.sty can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/subfig/




% *** FLOAT PACKAGES ***
%
%\usepackage{fixltx2e}
% fixltx2e, the successor to the earlier fix2col.sty, was written by
% Frank Mittelbach and David Carlisle. This package corrects a few problems
% in the LaTeX2e kernel, the most notable of which is that in current
% LaTeX2e releases, the ordering of single and double column floats is not
% guaranteed to be preserved. Thus, an unpatched LaTeX2e can allow a
% single column figure to be placed prior to an earlier double column
% figure. The latest version and documentation can be found at:
% http://www.ctan.org/tex-archive/macros/latex/base/


%\usepackage{stfloats}
% stfloats.sty was written by Sigitas Tolusis. This package gives LaTeX2e
% the ability to do double column floats at the bottom of the page as well
% as the top. (e.g., "\begin{figure*}[!b]" is not normally possible in
% LaTeX2e). It also provides a command:
%\fnbelowfloat
% to enable the placement of footnotes below bottom floats (the standard
% LaTeX2e kernel puts them above bottom floats). This is an invasive package
% which rewrites many portions of the LaTeX2e float routines. It may not work
% with other packages that modify the LaTeX2e float routines. The latest
% version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/sttools/
% Do not use the stfloats baselinefloat ability as IEEE does not allow
% \baselineskip to stretch. Authors submitting work to the IEEE should note
% that IEEE rarely uses double column equations and that authors should try
% to avoid such use. Do not be tempted to use the cuted.sty or midfloat.sty
% packages (also by Sigitas Tolusis) as IEEE does not format its papers in
% such ways.
% Do not attempt to use stfloats with fixltx2e as they are incompatible.
% Instead, use Morten Hogholm'a dblfloatfix which combines the features
% of both fixltx2e and stfloats:
%
% \usepackage{dblfloatfix}
% The latest version can be found at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/dblfloatfix/




%\ifCLASSOPTIONcaptionsoff
%  \usepackage[nomarkers]{endfloat}
% \let\MYoriglatexcaption\caption
% \renewcommand{\caption}[2][\relax]{\MYoriglatexcaption[#2]{#2}}
%\fi
% endfloat.sty was written by James Darrell McCauley, Jeff Goldberg and 
% Axel Sommerfeldt. This package may be useful when used in conjunction with 
% IEEEtran.cls'  captionsoff option. Some IEEE journals/societies require that
% submissions have lists of figures/tables at the end of the paper and that
% figures/tables without any captions are placed on a page by themselves at
% the end of the document. If needed, the draftcls IEEEtran class option or
% \CLASSINPUTbaselinestretch interface can be used to increase the line
% spacing as well. Be sure and use the nomarkers option of endfloat to
% prevent endfloat from "marking" where the figures would have been placed
% in the text. The two hack lines of code above are a slight modification of
% that suggested by in the endfloat docs (section 8.4.1) to ensure that
% the full captions always appear in the list of figures/tables - even if
% the user used the short optional argument of \caption[]{}.
% IEEE papers do not typically make use of \caption[]'s optional argument,
% so this should not be an issue. A similar trick can be used to disable
% captions of packages such as subfig.sty that lack options to turn off
% the subcaptions:
% For subfig.sty:
% \let\MYorigsubfloat\subfloat
% \renewcommand{\subfloat}[2][\relax]{\MYorigsubfloat[]{#2}}
% However, the above trick will not work if both optional arguments of
% the \subfloat command are used. Furthermore, there needs to be a
% description of each subfigure *somewhere* and endfloat does not add
% subfigure captions to its list of figures. Thus, the best approach is to
% avoid the use of subfigure captions (many IEEE journals avoid them anyway)
% and instead reference/explain all the subfigures within the main caption.
% The latest version of endfloat.sty and its documentation can obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/endfloat/
%
% The IEEEtran \ifCLASSOPTIONcaptionsoff conditional can also be used
% later in the document, say, to conditionally put the References on a 
% page by themselves.




% *** PDF, URL AND HYPERLINK PACKAGES ***
%
%\usepackage{url}
% url.sty was written by Donald Arseneau. It provides better support for
% handling and breaking URLs. url.sty is already installed on most LaTeX
% systems. The latest version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/url/
% Basically, \url{my_url_here}.




% *** Do not adjust lengths that control margins, column widths, etc. ***
% *** Do not use packages that alter fonts (such as pslatex).         ***
% There should be no need to do such things with IEEEtran.cls V1.6 and later.
% (Unless specifically asked to do so by the journal or conference you plan
% to submit to, of course. )


% correct bad hyphenation here
%\hyphenation{op-tical net-works semi-conduc-tor}

\usepackage{verbatim}
\begin{document}
%
% paper title
% can use linebreaks \\ within to get better formatting as desired
% Do not put math or special symbols in the title.
\title{Analysis of Cartesian Genetic Programming's Evolutionary Mechanisms}
%
%
% author names and IEEE memberships
% note positions of commas and nonbreaking spaces ( ~ ) LaTeX will not break
% a structure at a ~ so this keeps an author's name from being broken across
% two lines.
% use \thanks{} to gain access to the first footnote area
% a separate \thanks must be used for each paragraph as LaTeX2e's \thanks
% was not built to handle multiple paragraphs
%

%\author{Michael~Shell,~\IEEEmembership{Member,~IEEE,}
%        John~Doe,~\IEEEmembership{Fellow,~OSA,}
%        and~Jane~Doe,~\IEEEmembership{Life~Fellow,~IEEE}% <-this % stops a space
\author{Brian~W.~Goldman, William~F.~Punch}
%\thanks{M. Shell is with the Department
%of Electrical and Computer Engineering, Georgia Institute of Technology, Atlanta,
%GA, 30332 USA e-mail: (see http://www.michaelshell.org/contact.html).}% <-this % stops a space
%\thanks{J. Doe and J. Doe are with Anonymous University.}% <-this % stops a space
%\thanks{Manuscript received April 19, 2005; revised December 27, 2012.}}

% note the % following the last \IEEEmembership and also \thanks - 
% these prevent an unwanted space from occurring between the last author name
% and the end of the author line. i.e., if you had this:
% 
% \author{....lastname \thanks{...} \thanks{...} }
%                     ^------------^------------^----Do not want these spaces!
%
% a space would be appended to the last name and could cause every name on that
% line to be shifted left slightly. This is one of those "LaTeX things". For
% instance, "\textbf{A} \textbf{B}" will typeset as "A B" not "AB". To get
% "AB" then you have to do: "\textbf{A}\textbf{B}"
% \thanks is no different in this regard, so shield the last } of each \thanks
% that ends a line with a % and do not let a space in before the next \thanks.
% Spaces after \IEEEmembership other than the last one are OK (and needed) as
% you are supposed to have spaces between the names. For what it is worth,
% this is a minor point as most people would not even notice if the said evil
% space somehow managed to creep in.



% The paper headers
\markboth{IEEE Transactions on Evolutionary Computation,~Vol.~\#\#, No.~\#, Month~Year}%
{IEEE Transactions on Evolutionary Computation,~Vol.~\#\#, No.~\#, Month~Year}
% The only time the second header will appear is for the odd numbered pages
% after the title page when using the twoside option.
% 
% *** Note that you probably will NOT want to include the author's ***
% *** name in the headers of peer review papers.                   ***
% You can use \ifCLASSOPTIONpeerreview for conditional compilation here if
% you desire.




% If you want to put a publisher's ID mark on the page you can do it like
% this:
%\IEEEpubid{0000--0000/00\$00.00~\copyright~2012 IEEE}
% Remember, if you use this you must call \IEEEpubidadjcol in the second
% column for its text to clear the IEEEpubid mark.



% use for special paper notices
%\IEEEspecialpapernotice{(Invited Paper)}




% make the title area
\maketitle

% As a general rule, do not put math, special symbols or citations
% in the abstract or keywords.
\begin{abstract}
The abstract goes here.
\end{abstract}

% Note that keywords are not normally used for peerreview papers.
\begin{IEEEkeywords}
TODO Keywords
\end{IEEEkeywords}






% For peer review papers, you can put extra information on the cover
% page as needed:
% \ifCLASSOPTIONpeerreview
% \begin{center} \bfseries EDICS Category: 3-BBND \end{center}
% \fi
%
% For peerreview papers, this IEEEtran command inserts a page break and
% creates the second title. It will be ignored for other modes.
\IEEEpeerreviewmaketitle



\section{Introduction}
% The very first letter is a 2 line initial drop letter followed
% by the rest of the first word in caps.
% 
% form to use if the first word consists of a single letter:
% \IEEEPARstart{A}{demo} file is ....
% 
% form to use if you need the single drop letter followed by
% normal text (unknown if ever used by IEEE):
% \IEEEPARstart{A}{}demo file is ....
% 
% Some journals put the first two words in caps:
% \IEEEPARstart{T}{his demo} file is ....
% 
% Here we have the typical use of a "T" for an initial drop letter
% and "HIS" in caps to complete the first word.
\IEEEPARstart{T}{ODO this} section.
% You must have at least 2 lines in the paragraph with the drop letter
% (should never be an issue)

\section{Cartesian Genetic Programming}
TODO Explain basic CGP, use~\cite{miller:2011:chapter2}.

\section{Duplicate Evaluation Avoidance}

TODO Explain here why only compared with parent, not global history.

\subsection{Skip}
The \emph{Skip} method for avoiding duplicate evaluations involves the least
amount of modification from CGP's normal behavior.  After an offspring is produced
and its set of active nodes is determined but before it is evaluated, it is compared
with its parent.  Each gene in each active node is compared for equivalence with
the corresponding gene in the offspring's parent.  If all genes are found to be equal
than the individual is not evaluated.  Instead the offspring is given the same fitness
as its parent, as there is no way for it to be phenotypically different from its parent.
Note that since inactive genes are not compared for equivalence it is still possible
for offspring to genetically differ from their parents.

Because \emph{Skip} does not modify any evolutionary mechanisms from normal CGP,
in all possible cases it can do no worse.  If the mutation rate is high enough relative
to the number of active genes, \emph{Skip} will act identically to normal CGP as
the probability of an offspring being actively identical to its parent is effectively zero.
When the mutation rate is low enough relative to the number of active genes, some number of
offspring will be actively identical to their parents, resulting in a reduction in evaluations
but no change in the evolutionary trajectory.

From a tuning perspective we would expect both intuitively and from initial
experimentation~\cite{goldman:2013:cgpwaste} that \emph{Skip} will be less sensitive
to the mutation rate and will perform best using a lower mutation rate than normal CGP.
The former comes from the fact that in normal CGP any mutation rate that has a significant
probability of creating offspring actively identical to their parents will be penalized
by wasting evaluations.  In \emph{Skip} there are only two penalties for reducing the
mutation rate.  First, if search becomes trapped in a local optima, low mutation
rates may have increased difficulty escaping to find the global optimum.  Second,
exceptionally low mutation rates may result in large computational overhead as
more and more time is spent attempting to produce an evaluable useful offspring.
Note these penalties also exist in normal CGP.

\subsection{Accumulate}
Similar to \emph{Skip}, \emph{Accumulate} works by adding a step between offspring
creation and offspring evaluation.  Instead of skipping evaluations with offspring
are determined to be actively identical to their parents, \emph{Accumulate} enters
into a cycle of repeated mutation until an individual worth evaluating is created.
While the working offspring is actively identical to the
parent, a new individual is produced by applying mutation to the working offspring.
Note that each mutation step can probabilistically mutate any number of genes, just
as in the mutation used by normal CGP.
In this way each iteration the working offspring accumulates mutations to inactive
genes until an individual is produced that differs by one or more active genes from
the parent.  This repeatedly mutated offspring is then evaluated.  If its fitness
is less than the original parent, the final set of mutations are reverted
and the offspring's fitness is set to the same as the original parent.

Viewed in another way, \emph{Accumulate} performs a micro evolution on each offspring
before it is fully produced.  The final result of each offspring step is to produce
an offspring which is either actively identical to its parent, but with mutations
to its inactive genes, or an offspring which is at least as fit as the parent but
not actively identical.

Previous experimentation has shown that \emph{Accumulate} acts very similarly to
\emph{Skip}~\cite{goldman:2013:cgpwaste}, with the exception that \emph{Accumulate}
favors lower mutation rates. While at first these algorithms may appear
quite different, further consideration shows how they are similar.  In \emph{Skip}
an actively identical offspring is likely to be selected as we expect mutations to
active genes to more often reduce fitness than in prove it.  In the next generation
this actively identical offspring is then mutated again, with its lineage likely
continuing until it finally does mutate an active gene.  At this point if the
mutant is better it replaces the parent, otherwise the parent, which has been accumulating
mutations to inactive genes, is kept.  In this way \emph{Skip} mirrors \emph{Accumulate},
except over multiple generations instead of compressed into a single generation.

\subsection{Single}
While \emph{Skip} and \emph{Accumulate} focus on what to do when offspring are
produced which are actively identical to their parents, \emph{Skip} changes
how mutation works to ensure only evaluable offspring are created.  Instead
of mutating each gene at a set probability, \emph{Single} chooses genes at
random to mutate until exactly one active gene is mutated.

This modification gives \emph{Single} three properties distinct from the other
forms of duplicate avoidance.  First, \emph{Single} forces offspring to have
an active gene which is different than there parent.  This limits drift as the
mutated active gene must either be to an intron or represent another way to
code a solution of equal quality.  Second, as a benefit of forced changes
\emph{Single} avoids the overhead of repeatedly generating individuals without
creating an evaluable offspring.  Third, \emph{Single} does not require a mutation
rate parameter, effectively setting the mutation rate to $\frac{1}{a}$ for active
genes and $\frac{1}{a+1}$ for inactive genes, where $a$ is the number of active genes.

In encodings without inactive genes, limiting mutation to changing exactly one gene
could prevent an algorithm from escaping some types of local maxima.  Yet because
CGP allows for inactive genes, \emph{Single} is still able to escape most local
maxima using sufficient drift of inactive genes.  When a high percentage of
the genome becomes active or when \emph{Single} is otherwise limited in its ability
to drift due to lack of introns, it does have an increased potential to become stuck.
As will be discussed in Section~\ref{sec-ordering} and was shown in~\cite{goldman:2013:ordering}
this problem of highly active genomes is very unlikely.

\section{Genome Ordering}
\label{sec-ordering}
TODO General stuff about ordering, including normal CGP.


\subsection{Reorder}
\label{sec:reorder}
The concept behind \emph{Reorder} is to randomly shuffled node ordering without
impacting gene expression.  This is possible because for a given node, there can
be a large number of other nodes which neither have no required ordering,
as they neither directly or indirectly take input from the node nor provide their
output to the node.  In general, this process works by assigning nodes new
locations in the genome at random once all of the nodes they take input from have been assigned
locations earlier in the genome.

The first step in performing \emph{Reorder} on a genome is to create convenience
data structures to store direct connection relationships, such that given a node
we can determine in constant time where it gets its input and which nodes read
its output.  These structures can be build in order $O(N)$ time, where $N$ is the
number of nodes.

With this preprocessing complete, the algorithm given in Figure~\ref{fig:reorder}
is used to assign each node a new location in the modified genome.
It starts by constructing the \emph{addable} set, which contains all nodes who's direct
dependencies have already been added.  Initially this contains only the input locations.
Once a node is assigned a location, all of the nodes that depend on its output
have that dependency marked as satisfied.  Once all of a nodes dependencies
are satisfied, it can be put into \emph{addable}.  In this way, nodes are
randomly removed from \emph{addable}, assigned the next possible location
in the genome, with new nodes added to \emph{addable} as it becomes semantically
viable.  Iteration ends when all nodes have been assigned new locations.
This algorithm requires $O(AN)$, where $A$ is the arity used by the nodes
and $N$ is the number of nodes.

\begin{figure}
  \begin{algorithmic}
    \Procedure{NewOrder}{}
    \State $addable \leftarrow input\_locations$
    \State $index \leftarrow |input\_locations|$
    \State $new\_loc \leftarrow \emptyset$
    \While{$|addable| > 0$}
      \State $working \leftarrow randpop(addable)$
      \If{$working \notin input\_locations$}
        \State $new\_loc[working] \leftarrow index$
        \State $index \leftarrow index + 1$
      \EndIf
      \ForAll{$link \in feeds\_to(working)$}
        \State Satisfy $link$'s dependence on $working$
        \If{$|unsatisfied(link)| = 0$}
          \State $addable \leftarrow addable \cup \{link\}$
        \EndIf
      \EndFor
    \EndWhile
    \State\Return $new\_loc$
    \EndProcedure
  \end{algorithmic}
  \caption{Algorithm used by \emph{Reorder} to determine new node locations.}
  \label{fig:reorder}
\end{figure}

The final step in \emph{Reorder} is to use the list of new locations to convert
the existing genome into the reordered genome.  This can be done in the straight
forward manner by converting connection genes and output location genes using
the $new_loc$ map returned from the algorithm in Figure~\ref{fig:reorder}.
This also requires $O(AN)$ time, as all connection genes in all nodes must
be converted.  As this is the final step unique to \emph{Reorder}, and
no previous steps were of higher complexity, it is also the complexity of
the algorithm as a whole.  Furthermore it is of the same complexity class
of copying the entire genome, meaning it does not change CGP's overall complexity.

\emph{Reorder} is used once each generation to shuffle the nodes of the parent
organism.  As the shuffling does not semantically change the parent, it does
not need to be reevaluated.  Furthermore, as shuffling does not change the requirement
that nodes only depend on those preceding them in the genome, shuffling does not
require any changes to any other CGP methods, such as evaluation and mutation.
Yet after shuffling, the potential mutations that can be applied to the parent
have changed.  Consider two nodes $X$ and $Y$ which are in the genome, but
have no dependence relationship.  As $X$ does not take input from $Y$ and
$Y$ does not take input from $X$, either can be placed preceding the other
in the genome.  Without shuffling, the ordering in which these two nodes originally
evolved will be preserved indefinitely, such that if $X$ precedes $Y$ at node creation,
$X$ will always precede $Y$.  No mutation in \emph{Normal} is capable of making $X$
take input from $Y$.  Yet CGP using \emph{Reorder} is capable of shuffling $Y$ to
precede $X$, allowing subsequent mutations to make the connection.

This node movement has the potential to reduce node reinvention.  If evolution created
$X$ and $Y$, and mutating $X$ to depend on $Y$ would improve fitness, mutation's
ability to make that improvement would depend on happenstance.  If $X$ evolved
preceding $Y$, mutation would have to recreate either $X$ or $Y$ in new locations
in order to make the required connection.  \emph{Reorder} could achieve the same
effect through shuffling, and without costly node reinvention.

The price \emph{Reorder} pays for this potential improvement is increased mutational search
space and minor runtime costs.  Even though \emph{Reorder} is not capable of representing
different individuals that \emph{Normal}, their mutational adjacency is significantly different.
By mutational adjacency, we mean the set of individual's which are most likely to be created
when mutating a given individual.  For instance in \emph{Normal} if $X$ precedes $Y$, individuals
containing $Y$ dependent on $X$ could be adjacent to the original individual, but those containing
$X$ dependent on $Y$ would not.  As \emph{Reorder} can shuffle node order, both sets would be
adjacent.  In this way ordering limits how \emph{Normal} performs its search.  This limitation
will benefit \emph{Normal} any time making $X$ dependent on $Y$ is detrimental, as
\emph{Normal} won't waste evaluations on individuals having that mutation.


\subsection{DAG}
The principle behind \emph{DAG} is to remove the traditional CGP requirement that
nodes can only take input from nodes that precede them in the genome.  Instead
this is replaced with the relaxed requirement than the genome cannot encode for
any cycles.  To accomplish this goal, modifications must be made to how CGP
performs mutation and evaluation.

When a connection gene is chosen for mutation, its value is still chosen randomly
from all possible alternative values.  To achieve this, we incrementally determine
which nodes transitively depend on the mutating node and which are not, with
information stored in the $known$ map.  Initially, we know that the node
is dependent on itself and that input locations and the nodes current inputs
are not dependent on the mutating node.  Nodes are then
tested in a random order for dependence using the algorithm given in Figure~\ref{fig:dag}.
The first node returned that is not dependent on the mutating node is then
used as the new connection gene value.

\begin{figure}
  \begin{algorithmic}
  \Procedure{IsDependent}{$working$, $known$}
    \If{$working \in known$}
      \State\Return{$known[working]$}
    \EndIf
    \ForAll{$link \in reads\_from(working)$}
      \If{\Call{IsDependent}{$link$, $known$}}
        \State $known[working] \leftarrow True$
        \State\Return{$True$}
      \EndIf
    \EndFor
    \State $known[working] \leftarrow False$
    \State\Return{$False$}
  \EndProcedure
  \end{algorithmic}
  \caption{Algorithm used by \emph{DAG} to determine valid connection gene values.}
  \label{fig:dag}
\end{figure}

The algorithm given in Figure~\ref{fig:dag} is a non-repeated recursive depth
first search of the individual's DAG.  Search terminates as soon as a dependent
node is found, and information from previous searches is recorded to prevent
repeated search.  Each recursion level results in a node having its dependency
set, meaning early calls will likely result in lots of recursion, but subsequent
calls will find answers faster.  In the worst case, this algorithm may need
to examine all nodes in the genome to determine if $working$ is dependent on the
mutating node, giving it a complexity of $O(N)$.  In the worst case, this algorithm
may be called on all possible nodes in the genome to check their dependence.
Yet cumulatively those calls can only take $O(N)$ time due to the non-repeating
nature of the search.  As a result, the expected worst case run time for \emph{DAG}
mutation is $O(mAN^2)$, as when $m$ is the mutation rate we expect $mAN$
connection genes to be mutated, each requiring a worst case of $N$ dependency checks.
This is in contrast to \emph{Normal} and \emph{Reorder} which require $O(mAN)$ to perform
mutation, as individual mutations can be performed in constant time.

The method for evaluating \emph{DAG} individuals is very similar to the method for
efficiently evaluating \emph{Normal} and \emph{Reorder} individuals.  Just as before,
preprocessing is done to determine which nodes are active, and only those nodes are
executed during evaluation.  Just as before the process begins from the output locations,
recursively following connection genes and marking nodes as active until the input locations
are reached.  Here the process diverges, as in \emph{DAG} we must determine not only the set of
nodes that are active, but the order in which those nodes should be executed.

The algorithm given in Figure~\ref{fig:reorder} can be reused to determine the order
in which nodes should be executed during evaluation.  Instead of using $new\_loc$
to reorder the genome, we can use this map to specify what order nodes should be
executed in.  For instance, if a node was given the new location of $X$, we know
that once all nodes at locations preceding $X$ have been evaluated, the node at $X$
can be evaluated.  Stripping the inactive nodes from $new\_loc$ and inverting the map
results in an efficient, valid ordering in which to execute the nodes.

The process of determining which nodes are active requires $O(AN)$ time, regardless
of how the node ordering is handled.  As was discussed in Section~\ref{sec:reorder},
getting the $new\_loc$ map requires $O(AN)$ time.  Finally, converting the map
into an ordering of the active nodes takes $O(N)$ time.  Each of these steps are
sequential, meaning \emph{DAG} does not increase the runtime complexity of evaluation
preprocessing.  Furthermore, once the preprocessing step is done, \emph{DAG} takes
identical time to evaluate than the other methods, regardless of the number of
input combinations the individual is evaluated on.

With these modifications, there is no longer any meaning to the position of a node
in the genome.  In \emph{Normal} a node near the input locations will always
have a highly restricted set of values.  More completely, the number of valid
values for a connection gene in \emph{Normal} is related to the number of nodes
preceding that gene, and independent of the number of nodes following that gene.
In \emph{DAG} a connection gene's valid values only depends on the current solution
represented by the individual, independent of the genes location in the genome.
As such, there are far less artificial limitations on how \emph{DAG} can modify
solutions.

While both reduce mutational limitations, \emph{DAG}'s changes are more broad that
\emph{Reorder}.  While both have the possibility to use mutation to connect any two
nodes such that no cycle is created, \emph{Reorder} still maintains a bias about
how often some types of connection can happen.  Consider again two nodes $X$ and $Y$,
such that neither is dependent on the other.  If $X$ is transitively dependent on very
few nodes, and $Y$ is transitively dependent on many nodes, in general \emph{Reorder}
will shuffle the genome such that $X$ is before $Y$.  This is true because the probability
that all of $X$'s dependencies are added at any point of building the new order is always
higher than for $Y$'s dependencies.  As a result we expect $X$ to be put into $addable$ sooner,
and as a result we expect it to be added sooner.  The greater the discrepancy in these
set sizes the less likely the order between the nodes will be reversed.
Conversely, \emph{DAG} has no such bias.  It is always possible to mutate $X$ to connect to $Y$.

The probability of connecting $X$ to $Y$ is also influenced by how many nodes transitively depend on each.
In this example, let us assume $X$ has more nodes transitively dependent on it than $Y$ has.
In \emph{Reorder}, $X$ would again appear closer to the start of the genome, this time because the number
of possible nodes preceding it is reduced.  \emph{DAG} is biased in the reverse, such that $X$ will
have a higher chance of connecting to $Y$ that $Y$ to $X$.  This is because $X$ can connect to fewer
nodes, so each mutation has a higher chance of choosing $Y$.

\section{Qualitative Comparisons}
While it is possible to theorize the effects of each CGP variant on search, and
how those effects may interact with each other, without empirical evidence it is
mostly speculation.  Therefore we propose to start by testing the effectiveness
of the variants and their combinations in problem solving.  Deeper analysis of
the causes of their effectiveness are given in Section~\ref{sec:analysis}.

All told we have proposed three new methods for avoiding duplicate evaluations
(\emph{Skip}, \emph{Accumulate}, and \emph{Single}) and two new methods for
how to deal with genome ordering (\emph{Reorder} and \emph{DAG}).  As these two
sets are non overlapping, and there is no intuitive groupings, we chose to try
all possible combinations.  As a control we included \emph{Normal} ordering,
as it provides insight to how each of the duplicate prevention techniques work
if applied alone.  Similar measures are not needed in testing the ordering techniques,
as \emph{Skip} has no impact on evolutionary search, and is a strict improvement
to using no duplication detection at all.

In total this creates 9 algorithmic configurations.  Of these previous work
has been done testing the duplication methods with \emph{Normal} ordering~\cite{goldman:2013:cgpwaste}
as well as analysis of all three ordering methods with \emph{Single}~\cite{goldman:2013:ordering}.
This means there are 4 novel combinations, beyond this being the first time a cross
comparison has been performed.  This is also the first time the ordering methods
have been rigorously tested, as much of the work in~\cite{goldman:2013:ordering}
was analysis oriented.

\subsection{Problem Set}
In order to provide test landscapes for evolution, we chose problems common to
CGP literature and previous work with the duplication prevention and ordering
techniques.  We have chosen 4 binary circuit problems, as the binary representation
allows for some of the precise analysis performed in Section~\ref{sec:analysis}.
For all problems, we used the function set \{and, or, nand, nor\}.
To cover a range of different binary problems, each of the four chosen have
different numbers of inputs, different number of outputs, and different levels
of difficulty.

The first problem, 3-bit parity, represents probably the most
common historic application for
CGP~\cite{yu:2001:neutrality,miller:2006:redundancy,walker:2008:cgpmodules}.
We include this problem purely for backward comparison, as we agree that in general
it is too simple of a problem to merit its own conclusions~\cite{white:2013:bggb}.
Yet as part of the group it may help yield understanding about how each variant
performs on simple problems.

As representatives of binary problems with varying input and output sizes, we
reuse the 4-16-bit Decode and 16-4-bit Encode proposed in~\cite{goldman:2013:cgpwaste}.
In these problems, the circuit evolved must either convert a 4-bit encoded integer
to a 1 on the corresponding output line (Decode), or take a 1 on one of the 16
input lines and convert it back into a 4-bit encoded integer (Encode).  These
problems share properties with the commonly used Multiplexer problem, but include
multiple output lines, and can be evaluated quicker as each only requires 16 possible
test points.

Finally we chose the 3-bit Multiply to represent hard binary problems,
as was suggested in~\cite{white:2013:bggb} and used
by~\cite{vassilev:2000:neutrality,miller:2006:redundancy,walker:2008:cgpmodules}.
This problem is very difficult by comparison to the other problems, and has
the largest number of test points.



TODO Consider adding 5-bit Parity or Equality, 11-bit Multiplexer.

\subsection{Parameter Setting}
In order to fairly compare each method, we should ensure proper parameter configuration.
This is to avoid the potential criticism that the parameter settings chosen benefit
a specific method more than others.  While tuning can lead to the alternative issue
of methods only being effective after extensive problem specific tuning, we set out to use
rough grained values and provide each technique with equal tuning time to alleviate that bias.

We focus on two parameter values in our configuration: mutation rate and genome size.
While there are other potential configurable parameters (population size, offspring size, etc),
we feel the CGP literature has converged on settings for those parameters (1, 4).
Furthermore, as duplication detection explicitly deals with the relationship between
mutation rate and number of active genes, and the number of active genes depends
on the genome size and ordering method, mutation rate and genome size seem to be
the most likely to impact results.

To set parameter values, we started by defining a grid of parameter values:
50, 100, 200, 500, 1000, 2000, 5000, 10000 for genome size and 0.05, 0.02, 0.01,
0.005, 0.002, 0.001, 0.0005, 0.0002 for mutation rate.  Note that here a genome
size of 50 means there are 50 nodes, each composed of multiple genes, plus the
required genes to specify output locations.  Also note that mutation rate of 0.05
is used here to mean each gene has a 5\% probability of being mutated in a single
application of mutation.  These values were chosen as they completely cover the
range of previously used parameter settings, with each value having at least
one paired value for the other parameter which is feasibly effective.
The grid is formed by trying all possible mutation rates
with all possible genome sizes, resulting in 64 potential configurations for each
of the algorithmic configurations.  Finally note that as \emph{Single} does not
use a mutation rate, only the eight configurations with different genome sizes were
used when \emph{Single} was employed.

To choose which parameter configuration to use with each algorithmic configuration
for each problem, we used an iterative process of performing runs, comparing means, removing configurations,
and repeating.  For each algorithm configuration and problem, we first run all
potential parameter configurations 5 times (320 runs if not using \emph{Single}, 40 otherwise).
The parameter configurations are then sorted based on their median.  The best half are then
each run 4 more times, with the best half of that group chosen.  In this way, 32 configurations
are run 5 times, 16 are run 9 times, 8 are run 13 times, 4 are run 17 times, 2 are run 21 times,
and 2 are run 25 times.  For configuration with single, all 8 parameter configurations are run 17 times
before any are removed.  This iterative reduction allows us to focus tuning on
those parameter configurations most likely to be effective, with more information
gathered before choosing between high quality configurations.  In total this means
3924 complete runs are required to set the parameter configuration for all
algorithm configurations for a single problem.

To ensure termination, all runs were limited to 10,000,000 evaluations, well beyond
the expected time to completion.  As the median was used to compare configurations,
and not the mean, selection should be insensitive to this limit unless a large
portion of runs are failing to optimize.

To prevent undue biasing in the final results, none of these tuning runs were
used for any other data analysis.  This helps reduce selection's effect on
result distribution, similar to how it is common to split training and testing
runs.  In this case, the runs used to set the parameter configuration are considered
the training runs for that configuration.  As a result, our final results were gather
from a completely independent set of runs to test the configurations general characteristics.
Each configuration was run 50 times to ensure statistical power.

\subsection{Results}
TODO Compare using a table the best parameters found for each configuration.
One table per problem.


\subsection{Scalability}
TODO Discuss using tuned parameters on different problem sizes and analyze
results.  Include O(N) table similar to previous paper.

\subsection{Generalization}
TODO Consider dropping this section due to computational limits. 

\section{Analysis of Evolutionary Mechanisms}
\label{sec:analysis}
Comparing how many evaluations each algorithm takes to solve a set of benchmark
problems only provides us with a limited amount of understanding about how
those algorithms actually function.  We propose here to look for further detail,
both to check on theorized capabilities of each variant and CGP in general, and
to look for previously hidden aspects of CGP.
To do so we have devised a number of novel metrics to examine what makes a
CGP run successful.

As a primary concept to our analysis, we determine the \emph{footprint} of nodes
in the genome.  What we mean by \emph{footprint} is a measure the semantic use
of a node.  For each input that is given to an individual, all of its nodes
will produce a single output.  These output values can be recorded an ordered.
As the domain in use in binary, the outputs can be viewed as a bit string, with
each bit the value that a node outputs when an individual is presented with a
specific input.  Similar recording can be done for other domains, but binary
allows for by far the simplest encoding.

The \emph{footprint} of a node is a complete description of its functionality.
Any two nodes with identical \emph{footprint} values have identical behavior,
even if their method for encoding that behavior is drastically different.  From
this perspective, we can view CGP evolution as attempting to evolve a node who's
\emph{footprint} matches the desired output \emph{footprint}.  Doing so allows
us to examine how the evolutionary operators change the behavior of nodes,
not just their gene values.

Utilizing \emph{footprint} analysis, we can extract the semantically useful
portions of an individual, a subset of the active nodes.  Consider two active
nodes with the same \emph{footprint}.  As CGP is a graph representation, only
one of these nodes is actually necessary.  Given one in the genome, any node that reads
from the other can have its connection changed without changing its behavior.

\begin{figure}
  \begin{algorithmic}
  \Procedure{RepeatedFootprint}{}
    %\State $reachable \leftarrow \{i \mapsto \{ footprint(i) \} | i \in input\_locations\}$
    \State $reachable \leftarrow \emptyset$
    \ForAll{$i \in input\_locations$}
      \State $reachable[i] \leftarrow footprint(i)$
    \EndFor
    \State $repeated \leftarrow \emptyset$
    \ForAll{$node \in active$}
      \State $working \leftarrow footprint(node)$
      \State $direct \leftarrow \emptyset$
      \ForAll{$link \in reads\_from(node)$}
        \State $direct \leftarrow direct \cup reachable[link]$
      \EndFor
      \If{$working \in direct$}
        \State $repeated \leftarrow repeated \cup \{working\}$
      \Else
        \State $direct \leftarrow direct \cup \{working\}$
      \EndIf
      \State $reachable[node] \leftarrow direct$
    \EndFor
    \State\Return{$repeated$}
  \EndProcedure
  \end{algorithmic}
  \caption{Algorithm to determine which nodes transitively depend on their
  own footprint.}
  \label{fig:repeated}
\end{figure}

\begin{figure}
  \begin{algorithmic}
  \Procedure{Simplify}{$p\_r$, $p\_c$, $p\_i$, $best$}
    \If{$|p\_r| = 0$}
      \State\Return{$p\_i$}
    \ElsIf{$|p\_i| + |p\_r| \geq |best|$}
      \State\Return{$best$}
    \EndIf
    \State $working, p\_a \leftarrow randpop(p\_r)$
    \State $covrd \leftarrow p\_c \cup {working}$
    \State $ansc \leftarrow p\_a \cup {working}$
    \ForAll{$node \in has\_footprint(working)$}
      \State $node\_valid \leftarrow True$
      \State $incd \leftarrow p\_i \cup {node}$
      \State $reqd \leftarrow p\_r$
      \ForAll{$link \in reads\_from(node)$}
        \State $foot \leftarrow footprint(link)$
        \If{$foot \in ansc$}
          \State $node\_valid \leftarrow False$
          \State \textbf{break}
        \ElsIf{$foot \notin covrd$}
          \State $reqd[foot] \leftarrow reqd[foot] \cup ansc$
        \EndIf
      \EndFor
      \If{$node\_valid$}
        \State $solution \leftarrow \Call{Simplify}{reqd, covrd, incd, best}$
        \If{$|solution| < |best|$}
          \State $best \leftarrow solution$
        \EndIf
      \EndIf
    \EndFor
    \State\Return{$best$}
  \EndProcedure
  \end{algorithmic}
  \caption{Recursive algorithm which determines the minimum set of nodes required
    to generate a set of required outputs.}
  \label{fig:simplify}
\end{figure}



TODO Explain simplify algorithm

\subsection{Inactive Nodes and Node Duplication}
TODO Look at how much of the genome is duplicated in both active and inactive

TODO Look at how much of the genome is unused.

\subsection{Node Reactivation and Active Node Mutation}
TODO For each node that was active, then inactive, then active again, see how its changed.

TODO For each node that is active, look at how successful mutations change its behavior.

\subsection{Source of High Variance}
TODO Look for general indicators of high evaluation runs.

TODO Look for footprint moving.

TODO Look for active node clustering.

TODO Look into how often children replace parents.


% An example of a floating figure using the graphicx package.
% Note that \label must occur AFTER (or within) \caption.
% For figures, \caption should occur after the \includegraphics.
% Note that IEEEtran v1.7 and later has special internal code that
% is designed to preserve the operation of \label within \caption
% even when the captionsoff option is in effect. However, because
% of issues like this, it may be the safest practice to put all your
% \label just after \caption rather than within \caption{}.
%
% Reminder: the "draftcls" or "draftclsnofoot", not "draft", class
% option should be used if it is desired that the figures are to be
% displayed while in draft mode.
%
%\begin{figure}[!t]
%\centering
%\includegraphics[width=2.5in]{myfigure}
% where an .eps filename suffix will be assumed under latex, 
% and a .pdf suffix will be assumed for pdflatex; or what has been declared
% via \DeclareGraphicsExtensions.
%\caption{Simulation Results.}
%\label{fig_sim}
%\end{figure}

% Note that IEEE typically puts floats only at the top, even when this
% results in a large percentage of a column being occupied by floats.


% An example of a double column floating figure using two subfigures.
% (The subfig.sty package must be loaded for this to work.)
% The subfigure \label commands are set within each subfloat command,
% and the \label for the overall figure must come after \caption.
% \hfil is used as a separator to get equal spacing.
% Watch out that the combined width of all the subfigures on a 
% line do not exceed the text width or a line break will occur.
%
%\begin{figure*}[!t]
%\centering
%\subfloat[Case I]{\includegraphics[width=2.5in]{box}%
%\label{fig_first_case}}
%\hfil
%\subfloat[Case II]{\includegraphics[width=2.5in]{box}%
%\label{fig_second_case}}
%\caption{Simulation results.}
%\label{fig_sim}
%\end{figure*}
%
% Note that often IEEE papers with subfigures do not employ subfigure
% captions (using the optional argument to \subfloat[]), but instead will
% reference/describe all of them (a), (b), etc., within the main caption.


% An example of a floating table. Note that, for IEEE style tables, the 
% \caption command should come BEFORE the table. Table text will default to
% \footnotesize as IEEE normally uses this smaller font for tables.
% The \label must come after \caption as always.
%
%\begin{table}[!t]
%% increase table row spacing, adjust to taste
%\renewcommand{\arraystretch}{1.3}
% if using array.sty, it might be a good idea to tweak the value of
% \extrarowheight as needed to properly center the text within the cells
%\caption{An Example of a Table}
%\label{table_example}
%\centering
%% Some packages, such as MDW tools, offer better commands for making tables
%% than the plain LaTeX2e tabular which is used here.
%\begin{tabular}{|c||c|}
%\hline
%One & Two\\
%\hline
%Three & Four\\
%\hline
%\end{tabular}
%\end{table}


% Note that IEEE does not put floats in the very first column - or typically
% anywhere on the first page for that matter. Also, in-text middle ("here")
% positioning is not used. Most IEEE journals use top floats exclusively.
% Note that, LaTeX2e, unlike IEEE journals, places footnotes above bottom
% floats. This can be corrected via the \fnbelowfloat command of the
% stfloats package.



\section{Conclusion}
The conclusion goes here.





% if have a single appendix:
%\appendix[Proof of the Zonklar Equations]
% or
%\appendix  % for no appendix heading
% do not use \section anymore after \appendix, only \section*
% is possibly needed

% use appendices with more than one appendix
% then use \section to start each appendix
% you must declare a \section before using any
% \subsection or using \label (\appendices by itself
% starts a section numbered zero.)
%


\appendices
\section{Example Appendix}
Appendix one text goes here.

% you can choose not to have a title for an appendix
% if you want by leaving the argument blank
\section{}
Appendix two text goes here.


% use section* for acknowledgement
\section*{Acknowledgment}


The authors would like to thank...


% Can use something like this to put references on a page
% by themselves when using endfloat and the captionsoff option.
\ifCLASSOPTIONcaptionsoff
  \newpage
\fi



% trigger a \newpage just before the given reference
% number - used to balance the columns on the last page
% adjust value as needed - may need to be readjusted if
% the document is modified later
%\IEEEtriggeratref{8}
% The "triggered" command can be changed if desired:
%\IEEEtriggercmd{\enlargethispage{-5in}}

% references section

% can use a bibliography generated by BibTeX as a .bbl file
% BibTeX documentation can be easily obtained at:
% http://www.ctan.org/tex-archive/biblio/bibtex/contrib/doc/
% The IEEEtran BibTeX style support page is at:
% http://www.michaelshell.org/tex/ieeetran/bibtex/
%\bibliographystyle{IEEEtran}
% argument is your BibTeX string definitions and bibliography database(s)
%\bibliography{IEEEabrv,../bib/paper}
%
% <OR> manually copy in the resultant .bbl file
% set second argument of \begin to the number of references
% (used to reserve space for the reference number labels box)
%\begin{thebibliography}{1}

%\bibitem{IEEEhowto:kopka}
%H.~Kopka and P.~W. Daly, \emph{A Guide to \LaTeX}, 3rd~ed.\hskip 1em plus
%  0.5em minus 0.4em\relax Harlow, England: Addison-Wesley, 1999.

%\end{thebibliography}

\bibliographystyle{IEEEtran}
\bibliography{IEEEabrv,../main}

% biography section
% 
% If you have an EPS/PDF photo (graphicx package needed) extra braces are
% needed around the contents of the optional argument to biography to prevent
% the LaTeX parser from getting confused when it sees the complicated
% \includegraphics command within an optional argument. (You could create
% your own custom macro containing the \includegraphics command to make things
% simpler here.)
%\begin{IEEEbiography}[{\includegraphics[width=1in,height=1.25in,clip,keepaspectratio]{mshell}}]{Michael Shell}
% or if you just want to reserve a space for a photo:

\begin{IEEEbiography}{Brian W. Goldman}
TODO
\end{IEEEbiography}

% if you will not have a photo at all:
\begin{IEEEbiographynophoto}{William F. Punch}
TODO
\end{IEEEbiographynophoto}

% insert where needed to balance the two columns on the last page with
% biographies
%\newpage

% You can push biographies down or up by placing
% a \vfill before or after them. The appropriate
% use of \vfill depends on what kind of text is
% on the last page and whether or not the columns
% are being equalized.

%\vfill

% Can be used to pull up biographies so that the bottom of the last one
% is flush with the other column.
%\enlargethispage{-5in}



% that's all folks
\end{document}


